File: csharp/src/code-smells/dispensables/DataClass.cs
namespace RefactoringAvanzado.CodeSmells.Dispensables;

public class UserRecord
{
    public string Id { get; set; }
    public string Name { get; set; }
    public string Email { get; set; }
    public DateTime CreatedAt { get; set; }

    public UserRecord(string id, string name, string email, DateTime createdAt)
    {
        Id = id;
        Name = name;
        Email = email;
        CreatedAt = createdAt;
    }
}

public class UserService
{
    public UserRecord CreateUser(string name, string email)
    {
        if (!email.Contains('@'))
        {
            throw new ArgumentException("Invalid email");
        }

        return new UserRecord(Guid.NewGuid().ToString(), name, email, DateTime.Now);
    }

    public void UpdateUserEmail(UserRecord user, string newEmail)
    {
        if (!newEmail.Contains('@'))
        {
            throw new ArgumentException("Invalid email");
        }
        user.Email = newEmail;
    }
}

public class UserReportGenerator
{
    public string GenerateUserSummary(UserRecord user)
    {
        return $"User {user.Name} ({user.Email}) created on {user.CreatedAt.ToShortDateString()}";
    }
}

public static class DataClass
{
    public static string DemoDataClass()
    {
        var service = new UserService();
        var report = new UserReportGenerator();
        var user = service.CreateUser("Lina", "lina@example.com");
        service.UpdateUserEmail(user, "lina+news@example.com");
        return report.GenerateUserSummary(user);
    }
}

File: typescript/src/code-smells/dispensables/data-class.ts
import {v4 as uuidv4} from 'uuid'

export class UserRecord {
  constructor(
    public id: string,
    public name: string,
    public email: string,
    public createdAt: Date,
  ) {
  }
}

class UserService {
  createUser(name: string, email: string): UserRecord {
    if (!email.includes('@')) {
      throw new Error('Invalid email')
    }

    return new UserRecord(uuidv4(), name, email, new Date())
  }

  updateUserEmail(user: UserRecord, newEmail: string): void {
    if (!newEmail.includes('@')) {
      throw new Error('Invalid email')
    }
    user.email = newEmail
  }
}

class UserReportGenerator {
  generateUserSummary(user: UserRecord): string {
    return `User ${user.name} (${user.email}) created on ${user.createdAt.toLocaleDateString()}`
  }
}

export function demoDataClass(): string {
  const service = new UserService()
  const report = new UserReportGenerator()
  const user = service.createUser('Lina', 'lina@example.com')
  service.updateUserEmail(user, 'lina+news@example.com')
  return report.generateUserSummary(user)
}
